Minor fixes

dipamsen opened the pull request:
- Fixed 'X' lab slot having extra hour (according to latest Central Time Table)
- check for `argv[1]`'s existence before access

On running the script there seems to be a huge diff on `data/data.json`, which seems to be largely due to order changes (along with few actual changes to the data as well). Is the order of slots, rooms random for some reason?

dipamsen committed:
fix: error in slot map

dipamsen committed:
fix: check existence of argv before access

proffapt commented:
Hi, @dipamsen thanks for the contribution!!

We were indeed reported some inconsistencies in the displayed time-table for some professors. I am attaching those incosistencies in this message so that you can verify that your fix actually fixes them as well. If this doesn't you may try to solve it (in this same or different PR). 

![image](https://github.com/user-attachments/assets/78dabadc-3cb9-40dd-a494-511dbd239142)
![image](https://github.com/user-attachments/assets/814196ad-0e82-4a87-b81d-404c4acb8130)

Look for EC60101.

cc: @harshkhandeparkar @Faizan2005 @Prasanthi-Peram @rohan-b-84

dipamsen committed:
fix: display all possible courses at a given slot

dipamsen committed:
remove unused index

dipamsen committed:
revert formatting changes

dipamsen commented:
@proffapt  Thanks!

I found out the problem in the above example is that when multiple courses occupy the same slot [1], only the first one found is shown in the frontend. I have changed it to allow more than one course to be present in one slot. (I have not modified any styles or structure of the data, just changed the function `getTimeSlotInfo` to return an array of courses rather than a single one.)

![image](https://github.com/user-attachments/assets/55fb8d49-8225-4bdc-89fe-0b683541a3aa)

[1]: This can happen because there are courses which are divided into sections, and one professor only teaches one section. However the ERP does not give any details about which professor teaches which section, so we are forced to show all possible slots where the class could be happening, even if the professor may not be present there (like seen above for EC29201)

proffapt commented:
Great work!

Can you verify this prof's details too? or share the timetable after your fixes?

![image](https://github.com/user-attachments/assets/4c0638e7-ed59-4a61-9925-fb357828eebf)
EC60013

Once we get this verified and get approval for your code changes, then this PR will be merged.

> [1]: This can happen because there are courses which are divided into sections, and one professor only teaches one section. However the ERP does not give any details about which professor teaches which section, so we are forced to show all possible slots where the class could be happening, even if the professor may not be present there (like seen above for EC29201)

Makes sense. I guess we can add a disclaimer or a collapsable tooltip which tells the user about this nuances.
cc: @Prasanthi-Peram @Faizan2005 @harshkhandeparkar @rohan-b-84

dipamsen commented:
@proffapt Sure thing!

![image](https://github.com/user-attachments/assets/aa32fe8c-7fc9-4e5d-845d-b427d52c3089)

dipamsen committed:
Merge branch 'master' into master

harshkhandeparkar commented on wimp/constants.py:
```diff
@@ -102,7 +102,7 @@
     "V3": [(3, 6), (3, 7), (4, 6)],
     "V4": [(3, 6), (3, 7), (4, 5), (4, 6)],
-    "X": [(2, 5), (2, 6), (2, 7), (2, 8)],
+    "X": [(2, 5), (2, 6), (2, 7)],
```
Doesn't the X slot have 4 hours [here](https://erp.iitkgp.ac.in/Acad/AUT_Time_Table_24_25.pdf)?

harshkhandeparkar replied:
Technically it has 4 hours but only 3 of those are given to the lab slot X. Are there any examples where the 4th hour is also used? And if yes, how is it mentioned?

dipamsen replied:
No, because as I understand it, just "X" refers to the Lab Slot X (at least from our data source, i.e. "Subject List with Slots" in the ERP), which is just 3 hours long (as are all lab slots). "X4" would refer to the four hour slot.

dipamsen replied:
If the fourth slot is used, it would be mentioned as "X44" (fourth slot of X4)

dipamsen replied:
Example of fourth hour of X4 being used

![image](https://github.com/user-attachments/assets/2f34d4a7-cb07-4d5f-8073-2964bda36ffd)

harshkhandeparkar replied:
We should handle that case as well. Any ideas on how that could be done?

rohan-b-84 replied:
> just "X" refers to the Lab Slot X (at least from our data source, i.e. "Subject List with Slots" in the ERP), which is just 3 hours long (as are all lab slots). "X4" would refer to the four hour slot

afaik, this is the case.
If only X is mentioned, then its the Lab Slot X
If its X with numbers, then its the courses, which can also (or only) have 4th 5-6pm slot.

rohan-b-84 replied:
> Any ideas on how that could be done?

I guess, we can "X" mark the first 3 hours, and  "X4n" should mark the corresponding 1hour slots ??

dipamsen replied:
@rohan-b-84 I think this is already handled? I havent looked at the code, but it seems to work?
eg:  (X4 slots - wednesday 2pm to 6pm)

![image](https://github.com/user-attachments/assets/12e6912f-f8e3-4575-95c4-987e2654ab7a)

Without having looked at the code, I believe Aij is handled as `Ai[j-1]`, because Ai always is a list of slots.

Also, this notation is not specific to X slot, eg. S33 refers to third hour of S3 slot.

harshkhandeparkar reviewed (changes requested):
Have you committed the Python changes?

harshkhandeparkar commented:
> On running the script there seems to be a huge diff on data/data.json, which seems to be largely due to order changes (along with few actual changes to the data as well). Is the order of slots, rooms random for some reason?

I think it is because the timetables are fetched in random order.

dipamsen commented:
> Have you committed the Python changes?

Yes, there are only two lines of change :) (the multiple courses per slot thing was purely a frontend change)

proffapt commented:
> @proffapt Sure thing!
> 
> ![image](https://private-user-images.githubusercontent.com/59444569/371755734-aa32fe8c-7fc9-4e5d-845d-b427d52c3089.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc1MjYzODMsIm5iZiI6MTcyNzUyNjA4MywicGF0aCI6Ii81OTQ0NDU2OS8zNzE3NTU3MzQtYWEzMmZlOGMtN2ZjOS00ZTVkLTg0NWQtYjQyN2Q1MmMzMDg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI4VDEyMjEyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4MjlkYjQ1Y2M4Mjg2YWIzMTAxMWQyMzExNGQzMzM2ZWQ3NjQ0ODI3YTUwZmIyMDc4ZmM5Mjc5MGNiMTI5NjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zF2H2o4ClDf_ZoCEmHGXpGHw0WRnS3erMSbyjFtNTB8)

The one who reported this inconsistency has verified the correctness of this new time table.
Logic seems working fine. We can now proceed to actual code review.

cc: @harshkhandeparkar @rohan-b-84

proffapt commented:
> > Have you committed the Python changes?
> 
> Yes, there are only two lines of change :) (the multiple courses per slot thing was purely a frontend change)

Not exactly. The frontend displays the timetable structure taking data from `data.json` and `incosistencies.json`. Your changes fixed issues in these files and that's what the frontend displayed. So it's fine.

dipamsen commented:
> Not exactly. The frontend displays the timetable structure taking data from data.json and incosistencies.json. Your changes fixed issues in these files and that's what the frontend displayed. So it's fine.

Well, the `data.json` did already include information about the multiple courses in a single slot. It lists each course for a given professor, and for each course what slots it is present in. So naturally if two courses are present in one slot, that information is already available in `data.json`.

The only bug was in the frontend, wherein for each slot, the code searched for the first course in the array that occupied that slot. (It failed to account for the fact that there can be multiple such courses that occupies this slot)

proffapt commented:
> > Not exactly. The frontend displays the timetable structure taking data from data.json and incosistencies.json. Your changes fixed issues in these files and that's what the frontend displayed. So it's fine.
> 
> Well, the `data.json` did already include information about the multiple courses in a single slot. It lists each course for a given professor, and for each course what slots it is present in. So naturally if two courses are present in one slot, that information is already available in `data.json`.
> 
> The only bug was in the frontend, wherein for each slot, the code searched for the first course in the array that occupied that slot. (It failed to account for the fact that there can be multiple such courses that occupies this slot)

Ok, makes sense.

harshkhandeparkar reviewed (commented):
LGTM. Waiting for @rohan-b-84's review.

rohan-b-84 commented:
LGTM.

I would say we should add some consistent formatting/linting, and make the output more deterministic... but that can be another issue

proffapt commented:
Fine then, merging the PR and can discuss the linting part in a new issue.
